File: bdp-govern/metadata/atlas-versions/atlas-3.0.0-SNAPSHOT/atlas-examples/sample-app/src/main/java/org/apache/atlas/examples/sampleapp/EntityExample.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.atlas.examples.sampleapp;

import org.apache.atlas.AtlasClientV2;
import org.apache.atlas.AtlasServiceException;
import org.apache.atlas.model.instance.AtlasClassification;
import org.apache.atlas.model.instance.AtlasEntity;
import org.apache.atlas.model.instance.AtlasEntity.AtlasEntityWithExtInfo;
import org.apache.atlas.model.instance.AtlasEntityHeader;
import org.apache.atlas.model.instance.AtlasStruct;
import org.apache.atlas.model.instance.EntityMutationResponse;
import org.apache.atlas.model.instance.EntityMutations;
import org.apache.atlas.type.AtlasTypeUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static java.util.Arrays.asList;
import static org.apache.atlas.examples.sampleapp.SampleAppConstants.*;
import static org.apache.atlas.type.AtlasTypeUtil.toAtlasRelatedObjectId;
import static org.apache.atlas.type.AtlasTypeUtil.toAtlasRelatedObjectIds;

public class EntityExample {
    private static final String DATABASE_NAME                = "employee_db_entity";
    private static final String TABLE_NAME                   = "employee_table_entity";
    private static final String PROCESS_NAME                 = "employee_process_entity";
    private static final String METADATA_NAMESPACE_SUFFIX    = "@cl1";
    private static final String MANAGED_TABLE                = "Managed";
    private static final String ATTR_NAME                    = "name";
    private static final String ATTR_DESCRIPTION             = "description";
    private static final String ATTR_QUALIFIED_NAME          = "qualifiedName";
    private static final String REFERENCEABLE_ATTRIBUTE_NAME = ATTR_QUALIFIED_NAME;
    private static final String COLUMN_TIME_ID               = "time_id";
    private static final String COLUMN_CUSTOMER_ID           = "customer_id";
    private static final String COLUMN_COMPANY_ID            = "company_id";

    private final AtlasClientV2     client;
    private       AtlasEntity       dbEntity;
    private       AtlasEntity       tableEntityUS;
    private       AtlasEntity       tableEntityCanada;
    private       AtlasEntityHeader loadProcess;

    EntityExample(AtlasClientV2 client) {
        this.client = client;
    }

    public void createEntities() throws Exception {
        if (dbEntity == null) {
            dbEntity = createDatabaseEntity(DATABASE_NAME);

            SampleApp.log("Created entity: typeName=" + dbEntity.getTypeName() + ", qualifiedName=" + dbEntity.getAttribute(ATTR_QUALIFIED_NAME) + ", guid=" + dbEntity.getGuid());
        }

        if (tableEntityCanada == null) {
            tableEntityCanada = createTableEntity(TABLE_NAME + "_CANADA");

            SampleApp.log("Created entity: typeName=" + tableEntityCanada.getTypeName() + ", qualifiedName=" + tableEntityCanada.getAttribute(ATTR_QUALIFIED_NAME) + ", guid=" + tableEntityCanada.getGuid());
        }

        if (tableEntityUS == null) {
            tableEntityUS = createTableEntity(TABLE_NAME + "_US");

            SampleApp.log("Created entity: typeName=" + tableEntityUS.getTypeName() + ", qualifiedName=" + tableEntityUS.getAttribute(ATTR_QUALIFIED_NAME) + ", guid=" + tableEntityUS.getGuid());
        }

        if (loadProcess == null) {
            loadProcess = createProcessEntity(PROCESS_NAME);

            SampleApp.log("Created entity: typeName=" + loadProcess.getTypeName() + ", qualifiedName=" + loadProcess.getAttribute(ATTR_QUALIFIED_NAME) + ", guid=" + loadProcess.getGuid());
        }
    }

    public AtlasEntity getTableEntity() {
        return tableEntityUS;
    }

    public void getEntityByGuid(String entityGuid) throws Exception {
        AtlasEntityWithExtInfo entity = client.getEntityByGuid(entityGuid);

        if (entity != null) {
            SampleApp.log("Retrieved entity with guid=" + entityGuid);
            SampleApp.log("  " + entity);
        }
    }

    public void deleteEntities() throws Exception {
        client.deleteEntityByGuid(loadProcess.getGuid());

        SampleApp.log("Deleted entity: guid=" + loadProcess.getGuid());

        List<String> entityGuids = Arrays.asList(tableEntityUS.getGuid(), tableEntityCanada.getGuid(), dbEntity.getGuid());

        client.deleteEntitiesByGuids(entityGuids);

        SampleApp.log("Deleted entities:");
        for (String entityGuid : entityGuids) {
            SampleApp.log("  guid=" + entityGuid);
        }
    }

    private AtlasEntity createTableEntity(String tableName) throws Exception {
        return createHiveTable(dbEntity, tableName, MANAGED_TABLE,
                Arrays.asList(createColumn(COLUMN_TIME_ID, "int", "time id"),
                              createColumn(COLUMN_CUSTOMER_ID, "int", "customer id", SampleAppConstants.PII_TAG),
                              createColumn(COLUMN_COMPANY_ID, "double", "company id", SampleAppConstants.FINANCE_TAG)),
                SampleAppConstants.METRIC_TAG);
    }

    private AtlasEntityHeader createProcessEntity(String processName) throws Exception {
        return createProcess(processName, "hive query for monthly avg salary", "user ETL",
                asList(tableEntityUS),
                asList(tableEntityCanada),
                "create table as select ", "plan", "id", "graph", SampleAppConstants.CLASSIFIED_TAG);
    }

    private AtlasEntityHeader createProcess(String name, String description, String user, List<AtlasEntity> inputs, List<AtlasEntity> outputs,
                                            String queryText, String queryPlan, String queryId, String queryGraph, String... classificationNames) throws Exception {

        AtlasEntity entity = new AtlasEntity(SampleAppConstants.PROCESS_TYPE);

        entity.setAttribute(ATTR_NAME, name);
        entity.setAttribute(REFERENCEABLE_ATTRIBUTE_NAME, name + METADATA_NAMESPACE_SUFFIX);
        entity.setAttribute(ATTR_DESCRIPTION, description);
        entity.setAttribute(ATTR_USERNAME, user);
        entity.setAttribute(ATTR_START_TIME, System.currentTimeMillis());
        entity.setAttribute(ATTR_END_TIME, System.currentTimeMillis() + 10000);
        entity.setAttribute(ATTR_QUERY_TEXT, queryText);
        entity.setAttribute(ATTR_QUERY_PLAN, queryPlan);
        entity.setAttribute(ATTR_QUERY_ID, queryId);
        entity.setAttribute(ATTR_QUERY_GRAPH, queryGraph);
        entity.setAttribute(ATTR_OPERATION_TYPE, "testOperation");

        entity.setRelationshipAttribute(ATTR_INPUTS, toAtlasRelatedObjectIds(inputs));
        entity.setRelationshipAttribute(ATTR_OUTPUTS, toAtlasRelatedObjectIds(outputs));

        entity.setClassifications(toAtlasClassifications(classificationNames));

        return createEntity(new AtlasEntityWithExtInfo(entity));
    }

    private AtlasEntity createColumn(String name, String dataType, String comment, String... classificationNames) {
        AtlasEntity ret = new AtlasEntity(SampleAppConstants.COLUMN_TYPE);

        ret.setAttribute(ATTR_NAME, name);
        ret.setAttribute(REFERENCEABLE_ATTRIBUTE_NAME, name + METADATA_NAMESPACE_SUFFIX);
        ret.setAttribute(ATTR_DATA_TYPE, dataType);
        ret.setAttribute(ATTR_COMMENT, comment);

        ret.setClassifications(toAtlasClassifications(classificationNames));

        return ret;
    }

    private List<AtlasClassification> toAtlasClassifications(String[] classificationNames) {
        List<AtlasClassification> ret = new ArrayList<>();

        if (classificationNames != null) {
            for (String classificationName : classificationNames) {
                ret.add(new AtlasClassification(classificationName));
            }
        }

        return ret;
    }

    private AtlasEntityHeader createEntity(AtlasEntityWithExtInfo atlasEntityWithExtInfo) {
        EntityMutationResponse entity;

        try {
            entity = client.createEntity(atlasEntityWithExtInfo);

            if (entity != null && entity.getEntitiesByOperation(EntityMutations.EntityOperation.CREATE) != null) {
                List<AtlasEntityHeader> list = entity.getEntitiesByOperation(EntityMutations.EntityOperation.CREATE);

                if (list.size() > 0) {
                    return entity.getEntitiesByOperation(EntityMutations.EntityOperation.CREATE).get(0);
                }
            }
        } catch (AtlasServiceException e) {
            SampleApp.log("failed in create entity");
            e.printStackTrace();
        }

        return null;
    }

    private AtlasEntity createDatabaseEntity(String dbName) {
        AtlasEntity       hiveDBInstance = createHiveDBInstance(dbName);
        AtlasEntityHeader entityHeader   = createEntity(new AtlasEntityWithExtInfo(hiveDBInstance));

        if (entityHeader != null && entityHeader.getGuid() != null) {
            hiveDBInstance.setGuid(entityHeader.getGuid());
        }

        return hiveDBInstance;
    }

    protected AtlasEntity createHiveDBInstance(String dbName) {
        AtlasEntity entity = new AtlasEntity(SampleAppConstants.DATABASE_TYPE);

        entity.setAttribute(ATTR_NAME, dbName);
        entity.setAttribute(ATTR_DESCRIPTION, "employee database");
        entity.setAttribute(METADATA_NAMESPACE_SUFFIX, "employeeCluster");
        entity.setAttribute(REFERENCEABLE_ATTRIBUTE_NAME, dbName + METADATA_NAMESPACE_SUFFIX);
        entity.setAttribute(ATTR_OWNER, "user");
        entity.setAttribute(ATTR_LOCATION_URI, "/tmp");
        entity.setAttribute(ATTR_CREATE_TIME, 1000);

        return entity;
    }

    private AtlasEntity createHiveTable(AtlasEntity database, String tableName, String tableType, List<AtlasEntity> columns, String... classificationNames) throws Exception {
        AtlasEntityWithExtInfo entityWithExtInfo = new AtlasEntityWithExtInfo();

        AtlasEntity hiveTableInstance = createHiveTable(database, tableName, tableType, classificationNames);
        entityWithExtInfo.setEntity(hiveTableInstance);
        hiveTableInstance.setRelationshipAttribute(ATTR_COLUMNS, toAtlasRelatedObjectIds(columns));

        for (AtlasEntity column : columns) {
            column.setRelationshipAttribute(ATTR_TABLE, toAtlasRelatedObjectId(hiveTableInstance));
            entityWithExtInfo.addReferredEntity(column);
        }

        AtlasEntityHeader createdHeader = createEntity(entityWithExtInfo);

        if (createdHeader != null && createdHeader.getGuid() != null) {
            hiveTableInstance.setGuid(createdHeader.getGuid());
        }

        return hiveTableInstance;
    }

    private AtlasEntity createHiveTable(AtlasEntity database, String tableName, String tableType, String... classificationNames) throws Exception {
        AtlasEntity table = new AtlasEntity(SampleAppConstants.TABLE_TYPE);

        table.setAttribute(ATTR_NAME, tableName);
        table.setAttribute(REFERENCEABLE_ATTRIBUTE_NAME, database.getAttribute(ATTR_NAME) + "." + tableName + METADATA_NAMESPACE_SUFFIX);
        table.setAttribute(ATTR_TABLE_TYPE, tableType);
        table.setRelationshipAttribute(ATTR_DB, AtlasTypeUtil.getAtlasRelatedObjectId(database, TABLE_DATABASE_TYPE));

        table.setAttribute(ATTR_DESCRIPTION, "emp table");
        table.setAttribute(ATTR_LAST_ACCESS_TIME, "2014-07-11T08:00:00.000Z");
        table.setAttribute(ATTR_LEVEL, 2);
        table.setAttribute(ATTR_COMPRESSED, false);
        table.setClassifications(toAtlasClassifications(classificationNames));

        AtlasStruct serde1 = new AtlasStruct(STRUCT_TYPE_SERDE);

        serde1.setAttribute(ATTR_NAME, "serde1");
        serde1.setAttribute(ATTR_SERDE, "serde1");
        table.setAttribute(ATTR_SERDE1, serde1);

        AtlasStruct serde2 = new AtlasStruct(STRUCT_TYPE_SERDE);
        serde2.setAttribute(ATTR_NAME, "serde2");
        serde2.setAttribute(ATTR_SERDE, "serde2");
        table.setAttribute(ATTR_SERDE2, serde2);

        return table;
    }
}

File: govern/data-meta/atlas/atlas-examples/sample-app/src/main/python/entity_example.py
#!/usr/bin/env/python

#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import logging

from apache_atlas.model.enums    import EntityOperation
from apache_atlas.model.instance import AtlasEntityWithExtInfo, AtlasRelatedObjectId
from apache_atlas.utils          import type_coerce


LOG = logging.getLogger('entity-example')


class EntityExample:
    DATABASE_NAME                = "employee_db_entity"
    TABLE_NAME                   = "employee_table_entity"
    PROCESS_NAME                 = "employee_process_entity"
    METADATA_NAMESPACE_SUFFIX    = "@cl1"
    MANAGED_TABLE                = "Managed"
    ATTR_NAME                    = "name"
    ATTR_DESCRIPTION             = "description"
    ATTR_QUALIFIED_NAME          = "qualifiedName"
    REFERENCEABLE_ATTRIBUTE_NAME = ATTR_QUALIFIED_NAME
    ATTR_TIME_ID_COLUMN          = "time_id"
    ATTR_CUSTOMER_ID_COLUMN      = "customer_id"
    ATTR_COMPANY_ID_COLUMN       = "company_id"

    def __init__(self, client):
        self.client              = client
        self.entity_db           = None
        self.entity_table_us     = None
        self.entity_table_canada = None
        self.entity_process      = None

    def create_entities(self):
        self.__create_db()
        self.__create_us_table()
        self.__create_canada_table()
        self.__create_process()

    def get_table_entity(self):
        if self.entity_table_us:
            return self.entity_table_us

        return None

    def get_entity_by_guid(self, guid):
        entity = self.client.entity.get_entity_by_guid(guid)

        LOG.info("Entity(guid=%s): typeName=%s, attr.name=%s", guid, entity.entity.typeName, entity.entity.attributes['name'])

    def remove_entities(self):
        entity_list = [ self.entity_process.guid, self.entity_table_us.guid, self.entity_table_canada.guid, self.entity_db.guid ]

        self.client.entity.delete_entities_by_guids(entity_list)

        response = self.client.entity.purge_entities_by_guids(entity_list)

        if response is not None:
            LOG.info("Purged entities")
        else:
            LOG.info("Purge failed!")

    def __create_db(self):
        if not self.entity_db:
            with open('request_json/entity_create_db.json') as f:
                entity = type_coerce(json.load(f), AtlasEntityWithExtInfo)

                self.entity_db = self.__create_db_helper(entity)

        if self.entity_db:
            LOG.info("Created database entity: guid=%s, attr.name=%s", self.entity_db.guid, self.entity_db.attributes['name'])
        else:
            LOG.info("Failed to create database entity")

    def __create_us_table(self):
        if not self.entity_table_us:
            with open('request_json/entity_create_table_us.json') as f:
                entity = type_coerce(json.load(f), AtlasEntityWithExtInfo)

                self.entity_table_us = self.__create_table_helper(entity)

            if self.entity_table_us:
                LOG.info("Created US table entity: guid=%s, attr.name=%s", self.entity_table_us.guid, self.entity_table_us.attributes['name'])
            else:
                LOG.info("Failed to create US table entity")

    def __create_canada_table(self):
        if not self.entity_table_canada:
            with open('request_json/entity_create_table_canada.json') as f:
                entity = type_coerce(json.load(f), AtlasEntityWithExtInfo)

                self.entity_table_canada = self.__create_table_helper(entity)

            if self.entity_table_canada:
                LOG.info("Created Canada table entity: guid=%s, attr.name=%s", self.entity_table_canada.guid, self.entity_table_canada.attributes['name'])
            else:
                LOG.info("Failed to create Canada table entity")

    def __create_process(self):
        if not self.entity_process:
            with open('request_json/entity_create_process.json') as f:
                entity = type_coerce(json.load(f), AtlasEntityWithExtInfo)

                self.entity_process = self.__create_process_helper(entity)

        if self.entity_process:
            LOG.info("Created process entity: guid=%s, attr.name=%s", self.entity_process.guid, self.entity_process.attributes['name'])
        else:
            LOG.info("Failed to createa process entity")

    def __create_db_helper(self, entity):
        self.__create_entity(entity)

        return entity.entity

    def __create_table_helper(self, entity):
        table = entity.entity

        if self.entity_db:
            dbId = AtlasRelatedObjectId({ 'guid': self.entity_db.guid })

            LOG.info("setting: table(%s).db=%s", table.guid, dbId)

            table.relationshipAttributes['db'] = dbId

        self.__create_entity(entity)

        return table

    def __create_process_helper(self, entity):
        process = entity.entity

        process.relationshipAttributes = {}

        if self.entity_table_us:
            process.relationshipAttributes['inputs'] = [ AtlasRelatedObjectId({ 'guid': self.entity_table_us.guid }) ]

        if self.entity_table_canada:
            process.relationshipAttributes['outputs'] = [ AtlasRelatedObjectId({'guid': self.entity_table_canada.guid }) ]

        return self.__create_entity(entity)

    def __create_entity(self, entity):
        try:
            response = self.client.entity.create_entity(entity)

            guid = None

            if response and response.mutatedEntities:
                if EntityOperation.CREATE.name in response.mutatedEntities:
                    header_list = response.mutatedEntities[EntityOperation.CREATE.name]
                elif EntityOperation.UPDATE.name in response.mutatedEntities:
                    header_list = response.mutatedEntities[EntityOperation.UPDATE.name]

                if header_list and len(header_list) > 0:
                    guid = header_list[0].guid
            elif response and response.guidAssignments:
                if entity.entity is not None and entity.entity.guid is not None:
                    in_guid = entity.entity.guid
                else:
                    in_guid = None

                if in_guid and response.guidAssignments[in_guid]:
                    guid = response.guidAssignments[in_guid]

            if guid:
                entity.entity.guid = guid
        except Exception as e:
            LOG.exception("failed to create entity %s. error=%s", entity, e)

        return entity.entity if entity and entity.entity else None
